src/control/tools/CircleHandler.cpp

```diff
@@ -37,10 +39,14 @@ void CircleHandler::drawShape(Point& c, const PositionInputData& pos)
 	}
 	else
 	{
+		
 		double width = c.x - this->startPoint.x;
 		double height = c.y - this->startPoint.y;
+		
+		this->setModifiers(width, height,  pos);	// sets this->modShift and this->modControl
```

Comment:
For classes derived from BaseStrokeHandler:  This is pretty much all that's needed to use Shift and Control modifiers along with initial directional modifier. It sets this->modShift and this->modControl for use below. There's an optional boolean not used here to turn off cursor updates.